src/kv/map.h

```diff
@@ -205,101 +199,40 @@ namespace kv
       }
     }
 
-    // Serialisation-related methods
-    void serialise(KvStoreSerialiser& s, bool include_reads) override
+    // Used by owning map during serialise and deserialise
+    ChangeSet<K, V, H>& get_change_set()
     {
-      if (!changes)
-        return;
-
-      s.start_map(map.name, map.get_security_domain());
-
-      if (include_reads)
-      {
-        s.serialise_read_version(read_version);
-
-        s.serialise_count_header(reads.size());
-        for (auto it = reads.begin(); it != reads.end(); ++it)
-          s.serialise_read(it->first, it->second);
-      }
-      else
-      {
-        s.serialise_read_version(NoVersion);
-        s.serialise_count_header(0);
-      }
-
-      uint64_t write_ctr = 0;
-      uint64_t remove_ctr = 0;
-      for (auto it = writes.begin(); it != writes.end(); ++it)
-      {
-        if (!is_deleted(it->second.version))
-        {
-          ++write_ctr;
-        }
-        else
-        {
-          auto search = map.roll->get_tail()->state.get(it->first);
-          if (search.has_value())
-            ++remove_ctr;
-        }
-      }
-      s.serialise_count_header(write_ctr);
-      for (auto it = writes.begin(); it != writes.end(); ++it)
-      {
-        if (!is_deleted(it->second.version))
-        {
-          s.serialise_write(it->first, it->second.value);
-        }
-      }
-
-      s.serialise_count_header(remove_ctr);
-      for (auto it = writes.begin(); it != writes.end(); ++it)
-      {
-        if (is_deleted(it->second.version))
-        {
-          s.serialise_remove(it->first);
-        }
-      }
+      return change_set;
     }
 
-    bool deserialise(KvStoreDeserialiser& d, Version version) override
+    const ChangeSet<K, V, H>& get_change_set() const
     {
-      commit_version = version;
-      uint64_t ctr;
-
-      auto rv = d.template deserialise_read_version<Version>();
-      if (rv != NoVersion)
-        read_version = rv;
-
-      ctr = d.deserialise_read_header();
-      for (size_t i = 0; i < ctr; ++i)
-      {
-        auto r = d.template deserialise_read<K>();
-        reads[std::get<0>(r)] = std::get<1>(r);
-      }
-
-      ctr = d.deserialise_write_header();
-      for (size_t i = 0; i < ctr; ++i)
-      {
-        auto w = d.template deserialise_write<K, V>();
-        writes[std::get<0>(w)] = {0, std::get<1>(w)};
-      }
-
-      ctr = d.deserialise_remove_header();
-      for (size_t i = 0; i < ctr; ++i)
-      {
-        auto r = d.template deserialise_remove<K>();
-        writes[r] = {NoVersion, V()};
-      }
-
-      return true;
+      return change_set;
     }
 
-    bool is_replicated() override
+    void set_commit_version(Version v)
     {
-      return map.is_replicated();
+      commit_version = v;
     }
   };
 
+  template <typename K, typename V, typename H>
+  struct ConcreteTxView : public TxViewCommitter<K, V, H>,
+                          public TxView<K, V, H>
```

Comment:
May rename these for clarity.
- A `ChangeSet` is pure state: it holds some reads, writes, and versions, but has no logic for operating on them.
- A `TxView` lets us `get` and `put` changes, to interactively _modify_ a `ChangeSet`.
- A `TxViewCommitter` implements the `AbstractTxView` interface and deals with applying a `ChangeSet` to a `Map`.

This `ConcreteTxView` is both a `TxViewCommitter` (and transitively, `AbstractTxView`), and `TxView`: it can be used for either. But the _state_ it modifies is distinct, currently owned by `TxViewCommitter` and _referred_ to by `TxView` - it can also be modified by other things that look like a `TxView`.